### PR TITLE
Add collection conditions

### LINKED ISSUE #1372: Add new matcher syntax for comparing collections regardless of order
### Is your feature request related to a problem?

I'm always frustrated, when I want to make sure, that two collections have the same elements, regardless of their order. 

The current way to do it, is to cast both to `Set`, e.g., `x as Set == [1, 2, 3] as Set`, but that is verbose. Also if you want to make sure that duplicate Elements are recognized as well, then you'll have to use something like Hamcrest's [containsInAnyOrder](http://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/collection/IsIterableContainingInAnyOrder.html).

### Describe the solution you'd like

I'd like to be able to use a shorter Spock like syntax for comparing collections.
So, I propose to use `=~` and `==~` for this:

* `x =~ [1, 2, 3]` would be a lenient match, i.e., casting both to `Set` before comparing
* `x ==~ [1, 3, 3]` would be a strict match, using `containsInAnyOrder` under the hood so that duplicate elements are verified as well.

### Describe alternatives you've considered

_No response_

### Additional context

The proposed matchers have some limitations as we will have to do this in the AST transformation during compile time. So we only have static data available, and we don't want to break the existing behavior.

#### Limitations
* Would only be applied if either of the `Expression` in the `BinaryExpression`
  * Is a `VariableExpression` which `type` is a known `Collection` subclass or `Iterable`
  * Is a `ListExpression`

```groovy
  def "matchers"() {
    def x = [1]
    List y = [1]
    def z = [1]
    
    expect:
    x ==~ [1]  // works 
    x ==~ y    // works
    x ==~ z    // would default to normal Groovy Semantics as both are `Object` and fail for this case
  }
```

### PR DESCRIPTION
**Pull request by leonard84**
resolves #1372

### CODE REVIEW

File: docs/utilities.adoc [hunk 1]
```diff
@@ -37,3 +37,38 @@ include::{sourcedir}/utilities/MutableClockDocSpec.groovy[tag=age-filter-spec]
...
+[[collection-conditions]]
+== Collection Conditions
+
+Sometimes, you want to assert the contents of a collection regardless of its order.
+The Groovy way to do this is to cast both to `Set`, i.e. `x as Set == [1, 2, 3] as Set`.
+While this works, it is very noisy.
+
+Since Spock 2.1 you can use two new conditions:
+
+* `x =~ [1, 2, 3]` is a lenient match, i.e., casting both to `Set` before comparing.
```
Comment (marcphilipp) [inline docs/utilities.adoc:49]:
```suggestion
* `x =~ [1, 2, 3]` is a lenient match, i.e., checking that x contains at least one instance of every item in the list (same semantics as casting both to `Set` before comparing).
```

File: docs/utilities.adoc [hunk 2]
```diff
@@ -37,3 +37,38 @@ include::{sourcedir}/utilities/MutableClockDocSpec.groovy[tag=age-filter-spec]
...
+* `x ==~ [1, 2, 3, 3]` is a strict match, using Hamcrest's `containsInAnyOrder` under the hood so that duplicate elements are verified as well.
```
Comment (marcphilipp) [inline docs/utilities.adoc:50]:
```suggestion
* `x ==~ [1, 2, 3, 3]` is a strict match, i.e., checking that x contains exactly the items in the list regardless of their order (using Hamcrest's `containsInAnyOrder` under the hood).
```

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java [hunk 3]
```diff
@@ -201,4 +223,84 @@ static MatcherCondition parse(Object target, String method, Object[] args, boole
...
+
+    void verify(ErrorCollector errorCollector, @Nullable List<Object> values, @Nullable String text, int line, int column, @Nullable String message) {
+      String description = null;
+      int idxActual = values.indexOf(left);
+      int idxExpected = values.indexOf(right);
+
+      if (left instanceof Iterable && right instanceof Iterable) {
+        if (SpockRuntime.MATCH_COLLECTIONS_AS_SET.equals(method)) {
+          Set<?> actual = GroovyRuntimeUtil.coerce(left, LinkedHashSet.class);
+          Set<?> expected = GroovyRuntimeUtil.coerce(right, LinkedHashSet.class);
```
Comment (marcphilipp) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java:247]: Can we detect whether `right` was declared as a literal and fail if it had duplicate elements?

File: spock-core/src/main/java/org/spockframework/runtime/model/ExpressionInfo.java [hunk 4]
```diff
@@ -155,6 +155,15 @@ public void shiftVertically(int numLines) {
   public boolean isEqualityComparison() {
     return "==".equals(operation) && children.size() == 2;
   }
+  public boolean isSetComparison() {
```
Comment (marcphilipp) [inline spock-core/src/main/java/org/spockframework/runtime/model/ExpressionInfo.java:158]:
```suggestion

  public boolean isSetComparison() {
```

File: spock-core/src/main/java/org/spockframework/runtime/model/ExpressionInfo.java [hunk 5]
```diff
@@ -155,6 +155,15 @@ public void shiftVertically(int numLines) {
...
+    if("=~".equals(operation) && children.size() == 2) {
```
Comment (marcphilipp) [inline spock-core/src/main/java/org/spockframework/runtime/model/ExpressionInfo.java:159]:
```suggestion
    if ("=~".equals(operation) && children.size() == 2) {
```

**Review by marcphilipp — APPROVED** [review]
Looking forward to using that! 👍

File: docs/utilities.adoc [hunk 6]
```diff
@@ -37,3 +37,38 @@ include::{sourcedir}/utilities/MutableClockDocSpec.groovy[tag=age-filter-spec]
 
 There are many more ways to modify `MutableClock` just have a look at the JavaDocs, or the test code `spock.util.time.MutableClockSpec`.
 
+[[collection-conditions]]
+== Collection Conditions
+
+Sometimes, you want to assert the contents of a collection regardless of its order.
```
[the remark this answers is not in the document]
Comment (leonard84) [reply docs/utilities.adoc:43]: Both sound weird to me, WDYT about
```suggestion
Sometimes, you want to assert the elements of a collection regardless of their order.
```

[the remark this answers is not in the document]
Comment (marcphilipp) [reply docs/utilities.adoc:43]: SGTM

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java (hunk 3 above)
Comment (marcphilipp) [reply spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java:247]: It's probably unintended but I'm not sure whether we should fail in such cases. WDYT?

Comment (marcphilipp) [reply spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java:247]: Fair enough, let's not do it then.

**Review by marcphilipp — APPROVED** [review]

**Pull request merged**